Support NGSIv2 in Expression Translation, Attribute Alias and Event plugins

dcalvoalonso opened the pull request:
This PR aims to add NGSIv2 support to Expression Translation, Attribute Alias and Event plugins. The three plugins have been addressed at the same time since they are interconnected.

grunt test & grunt lint are passed.

dcalvoalonso committed:
NGSIv2 is supported in compress timestamp plugin

* ngsiService implements sending queries using NGSIv2

dcalvoalonso committed:
NGSIv2 support fot timestamp process plugin

dcalvoalonso committed:
Supporting NGSIv2 in: expressionPlugin and attributeAlias

dcalvoalonso committed:
Testing event plugins with NGSIv2

dcalvoalonso committed:
Merge branch 'task/activeAttsNgsi2' of https://github.com/dcalvoalonso/iotagent-node-lib into task/expressionPluginNgsi2

dcalvoalonso committed:
CHANGES_NEXT_RELEASE updated

dcalvoalonso committed:
Passing LINT tests

fgalan commented on lib/plugins/expressionParser.js:
```diff
@@ -163,6 +165,16 @@ function expressionApplier(context, typeInformation) {
 
         newAttribute.value = applyExpression(attribute.expression, context, typeInformation);
 
+        if (attribute.type === 'Integer') {
```
I understand this piece of code is due to different from NGSIv1, attribute values in NGSIv2 could use other native types in JSON appart from string. I see the numeric an boolean types covered, but, what about `null` (which is valid value for attribute in NGSIv2?

I'm not fully sury about it, maybe the comment doesn't apply... could you elaborate on the need of this piece of code, please?

dcalvoalonso replied:
You are completely right, I was missing some possible types: null, Object and Array.

In the last commit, I have considered these new three types and added the corresponding tests at the end of [expressionBasedTransformations-test.js](https://github.com/dcalvoalonso/iotagent-node-lib/blob/task/expressionPluginNgsi2/test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js)

dcalvoalonso replied:
Fixed in 0b20a6e218960a8a2732a3dbf78d7c4ac76b52a5

fgalan commented on test/unit/ngsiv2/examples/contextRequests/updateContextAliasPlugin.json:
```diff
@@ -0,0 +1,10 @@
+{
```
In this case, due to you have a updateContextAliasPlugin2.json file also, this one could be named updateContextAliasPlugin1.json

Not too important... you can close this comment with NTC if you think it doesn't worth the time to change.

fgalan replied:
(It seems this is the criteria you have followed with updateContextExpressionPlugin1.json and updateContextExpressionPlugin2.json)

dcalvoalonso replied:
Fixed in 0b20a6e218960a8a2732a3dbf78d7c4ac76b52a5

fgalan commented on test/unit/ngsiv2/examples/contextRequests/updateContextProcessTimestamp1.json:
```diff
@@ -0,0 +1,17 @@
+{
```
Is this file related with the PR? I mean, it mentions "process timestamp" plugin (that I think it was done in a previous PR). Moreover, the string updateContextExpressionPlugin1.json seems not being used in any test file in this PR...

dcalvoalonso replied:
Sorry for that! This file came somehow from previous PR.

dcalvoalonso replied:
Fixed in 0b20a6e218960a8a2732a3dbf78d7c4ac76b52a5

fgalan replied:
The werid thing is that if tha file was missing in the previous PR, that PR should have failed due to lack in this file... Could have a look to explain this little mistery, pls? :)

dcalvoalonso replied:
In the previous PR #2, originally I included for the unit tests a file called updateContextProcessTimestamp1.json. You suggested to use instead updateContextProcessTimestamp.json and I did the change.

The problem is that I had already started the work for this new PR #3 in a separate branch task/expressionPluginNgsi2 before you started providing your feedback to speed-up the following steps- Once you gave me the LGTM to PR#2, I merged the changes into in the new branch task/expressionPluginNgsi2 but the old updateContextProcessTimestamp1.json was not removed.

As you can see, the file was not missing in the previous PR because we changed its name. ;)

fgalan replied:
Ok, mistery clarified :)

NTC

fgalan commented on test/unit/ngsiv2/plugins/event-plugin_test.js:
```diff
@@ -0,0 +1,121 @@
+/*
```
I have seen in the code part of the PR modifications regarding Alias Attribute Plugin and Expressions Plugin but not related with Events plugin. Is this correct (maybe I have miss these modifications)?

In this case, I understands the modifications with regards Event plugin in this PR are the inclusion of a new NGSIv2-oriented test in order to check the functionally is still working, isn't it?

dcalvoalonso replied:
Yes, you are right, I haven't modified anything to support the functionality of Events plugin with NGSIv2, just added tests to demontrate that it works.

The reason is that this plugin just consists of the following lines:
```
function addEvents() {
    return (new Date()).toISOString();
}

exports.update = pluginUtils.createUpdateFilter(addEvents, config.getConfig().eventType || 'Event');
exports.query = pluginUtils.identityFilter;
```
And supporting NGSIv2 in pluginUtils was already addressed in the previous PR.

fgalan replied:
Ok. NTC.

dcalvoalonso committed:
Addressing comments of PR#3

fgalan commented on lib/plugins/expressionParser.js:
```diff
@@ -163,6 +165,22 @@ function expressionApplier(context, typeInformation) {
+            newAttribute.value = Number.parseFloat(newAttribute.value);
+        }
+        else if (attribute.type === 'Boolean') {
+            newAttribute.value = (newAttribute.value === 'true');
```
Just to be sure... is `'true`' or or shoud it be `true` ?

fgalan replied:
(by similarity with the `null` case just below)

dcalvoalonso replied:
Fixed in 1dfa28969796060f1306123c976449fa9fb9c48c

fgalan replied:
Maybe the test should cover the boolean case (and any other not yet covered now)?

dcalvoalonso replied:
@fgalan, I have analyzed in depth the functionality of this plugin to understand its expected results.

As it is explained in [expressionLanguage.md](https://github.com/telefonicaid/iotagent-node-lib/blob/master/doc/expressionLanguage.md), the objective of this plugin is: to adapt the information coming from the South Bound APIs to the information reported to the Context Broker by means of expressions.

As you can see in https://github.com/telefonicaid/iotagent-node-lib/blob/master/doc/expressionLanguage.md#types, it only supports two types: String or Number.  Also in this part of the documentation, it states: 
>For most arithmetic operations ('*', '/', etc...) if a variable is involved, its value will be cast to  Number, regardless of the original type. If this cast fails (because the value of the variable is not a number, e.g.: 'Fifty'), the overall result will be NaN.

So, if the plugin receives a Boolean, a NULL, a JSON Array or a JSON document, it first tries to convert it to a Number and to apply the expression. 

What happens with NGSIv1: as it only supports strings, for arithmetic expressions, the plugin tries to cast the value to a Number and calculates the expression. If it is not possible to do the conversion, it returns a NAN. If the expression is related to Strings, conversion is not needed. Here it is important to emphasize, that the plugin assumes that all the values that are going to come are strings as you can see in line 136 of https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js:
```
        result = parse(cleanedExpression, context, 'String');
```
The result of the expression will be a Number or a String depending on the type of expression and it is converted back to String in line 151 of this file.

Then, with NGSIv2, we have the following problems:
- Input values for expressions might be not only Strings and Numbers but also instances of the rest of JSON native types.
- Expression nevertheless, only supports Strings and Number.
There, I think that we should add an additional check so that expressions are never executed for unsupported types. With this, some of these lines that we have added would be not needed anymore since JSON arrays and object are not accepted by expressions.

Sorry for the long comment but I wanted to clarify everything before doing more changes.

What do you think?

fgalan replied:
It seems a very comprehensive analysis :) Let me some time to read carefully and think abou it...

fgalan replied:
> Input values for expressions might be not only Strings and Numbers but also instances of the rest of JSON native types.

Not sure if I'm getting the point... input values come from the southbound interface, not related with NGSIv2, and where the type semantics don't exist (as far as I know). All values are strings at southbound. Or maybe I'm wrong?

> Expression nevertheless, only supports Strings and Number.

In the northbound interface, I'd suggest to expand the interpretation of `type` in the following way:

* If type is "String", use the value as it is (no need to cast? is this the current behaviour with the NGSIv1 version of the plugin?)
* If type is "Number", cast the resulting value to JSON number (is this the current behaviour with the NGSIv1 version of the plugin?)
* If type is "Boolean", cast the resulting value to JSON boolean. We need a failover value in the case of casting error, it could be `false`
* If the type is "Null", use `null` as value. This is probably a bit useless, but I'd suggest to support it for compleness.


With regards to Arrays and Object, it would be cool to be able to do someething like this:

```
         "attributes":[
            ...
            {
               "name":"insideHouse",
               "type":"Array",
               "expression": [ "${@temperature}", "${@humidity}", "${@pressure}" ]
            },
            {
               "name":"permiter",
               "type":"Object",
               "expression": {
                 "north": "${@presenceN}",
                 "south": "${@presenceS}",
                 "west": "${@presenceW}",
                 "east": "${@presenceE}"
               }
            }
         ]
```
so, as result of the plugin, if we get a measure like this at southbound

```
t|25|h|79|p|27|pN|0|pS|1|pE|1|pW|0
```

the IOTA sends the following update to CB at northbound:

```
{
   "insideHouse": [ 25, 79, 27],
   "perimeter": {
     "north": 0,
     "south": 1,
     "east": 1,
     "west": 0
}
```

The array and object construct are possible in NGSIv1. Thus, if the current pluging doesn't support (could you confirm?) the above would be out the scope of the migration to NGSIv2. I other words, I'd suggest to focus only in Boolean and Null, but defer Array and Object support to a future enhancemente (and issues in the IOTAgents library repository should be filled about it).

(In fact, regarding Array and Objects, there are interesting challengues to solve... e.g. how to specify type in the sub-elements of the array or Object, i.e. `[ 25, 79, 27 ]` or `[ "25", "79", "27"]`. Maybe the cast could be specified in at expression level? i.e. `"expression": [ "(${@temperature},{Number}" ]`... just thinking aloud :)

The expressionLanguage.md document should be modified in order to take into account the new types and to explain any further NGSIv2 usage consideration to take into acount.


Does this makes any sense to you? I'm not fully sure (you would realize with the several in-line questions between brackets along the discussion above :), as I'm not an expert in this part of the code and the expression plugin is a bit tricky... :)

dcalvoalonso replied:
* Input values for expression plugin come from the southbound interface but before passing them as arguments for this plugin, they are converted to NGSI. Just to clarify everything, how is this plugin used?
  * First, in the corresponging IOTA, after calling to iotAgentLib.activate, you have to add this plugin:
```
        iotAgentLib.activate(iotAgentConfig, function() {
            iotAgentLib.clearAll(function() {
                iotAgentLib.addUpdateMiddleware(iotAgentLib.dataPlugins.attributeAlias.update);
                iotAgentLib.addQueryMiddleware(iotAgentLib.dataPlugins.attributeAlias.query);
                iotAgentLib.addUpdateMiddleware(iotAgentLib.dataPlugins.expressionTransformation.update);
                done();
            });
        });
```

  * In line 436 of [ngsiService.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/ngsiService.js), this simply implies that the plugin functionality is added to an array of middlewares to be called when an updateContext request is going to be sent to the CB as a result of having received active attributes from the southbound server.
```
function addUpdateMiddleware(middleware) {
    updateMiddleware.push(middleware);
}
```
  * In line 220 of [ngsiService.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/ngsiService.js), you will find the implementation of sendUpdateValue function.  In line 252 and using a NGSI payload as argument, the update middleware functions are called:
```
        apply(applyMiddlewares, updateMiddleware, payload, typeInformation)
```
* The NGSIv1 version of the plugin does not check any kind of type conversion for the values before applying the expression. It directly considers that the expression has String type: in line 131 of [expressionParser.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js):
```
function processExpression(context) {
    return function(expression) {
        var result,
            cleanedExpression = expression.substr(2, expression.length - 3);

        result = parse(cleanedExpression, context, 'String');

        return {
            original: expression,
            value: result
        };
    };
}
```
* It does not make either any kind of checking to the output values and directly updates the context information to be sent to the CB with the results of the expression. It uses replace instruction to the values will be always a string, which was OK for NGSIv1. In line 151 of the same file:
```
        expressionResult = expressionResult.replace(substitutions[i].original, substitutions[i].value);
```
* What happens then for NGSIv1 version if values which are not strings are received from the soutbound API:
  * If they are numbers: the expression parses seems to calculate correctly the expression and the result is converted to String (see previous point)
  * If they are booleans: the parser directly converts the input value to a number or string depending on the expression type and resolves the expression. 
  * If they are JSON objects or arrays: if the expression is of number type, the parser is not able to do a cast from these types to numbers and it returns NaN. This value is used to update the payload to be sent to the CB. If the expression is a string transformation (i.e., to trim spaces), it converts these values to strings and tries to resolve the operation.

dcalvoalonso replied:
So in order to maintain the scope of this PR and just to extend the plugin functionality to NGSIv2, I would follow what you are proposing:
* As happens with NGSIv1, don't do any kind of checking before applying the expressions.
* Try to cast to the appropriate types the results of the expressions. In NGSIv1, this cast was not explicitly (it was done with the replace instruction) needed since values for always represented using Strings. Strings, numbers, booleans and null types will be considered. JSON objects and arrays will be postponed and issues in the repository should be opened to analyze which changes are needed.

I think that we are aligned, aren't we? :)

fgalan replied:
Your proposal makes sense to me. Please, modify code and tests and let's see if all fits together and we have undestand each other. Maybe we don't achieve it in a first iteration, but I have the feeling we are progressing :)

dcalvoalonso replied:
Fixed in b217c46b403ab01d6dbbd1290d8e9db4c213e928.

So far, expressions only support the following values: string, integer, float, boolean and null. Therefore, the result of an expression is casted to the corresponding type. JSON arrays and objects are not supported yet so the behaviour will be the same that with NGSIv1.

I have added the following unit tests:
- Update for an integer attribute without expression
- Update for an integer attribute with expression
- Update for a float attribute without expression
- Update for a float attribute with expression
- Update for a Null attribute without expression
- Update for a Null attribute with arithmetic expression
- Update for a Null attribute with string expression (trim spaces)
- Update for a Boolean attribute without expression
- Update for a Boolean attribute with arithmetic expression
- Update for a Boolean attribute with string expression (trim spaces)
- Update for a JSON array without expression
- Update for a JSON object without expression

dcalvoalonso replied:
Please find attached coverage report 
![2017-11-29-153631_1366x716_scrot](https://user-images.githubusercontent.com/11388886/33380407-45a70c82-d51b-11e7-8038-3e1330352330.png)
for that function.

dcalvoalonso replied:
BTW, unit & lint tests are passed. :)

fgalan replied:
> I have added the following unit tests:
> [...]

I'm trying to map the cases in your bullet list with the ones in expressionBasedTransformations-test.js but I'm not fully sure... Could you annotate with comments in the .js file, e.g. something like this:

```
 describe('When an update comes for expressions with syntax errors', function() {
   // Case: Update for an ..
```

Thanks!

fgalan replied:
Btw, array/object functionality deferred to future issue https://github.com/telefonicaid/iotagent-node-lib/issues/568. Please provide any feedback you think could be usesful there.

dcalvoalonso replied:
Fixed in 895d5cc7b47d5fcfd4b81ab3e83a8a858c28c3e7

dcalvoalonso committed:
Address comments of PR#3

dcalvoalonso committed:
Address comments of PR #3

- More unit tests to cover applying expressions to Null and Boolean types
- JSON arrays and documents are not supported in the plugin yet since further analysis is required

fgalan commented on test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js:
```diff
@@ -0,0 +1,612 @@
+            {
+                name: 'p',
+                type: 'Integer',
+                value: 52
```
Just to check if I'm understanding correctly the "separtion of concerns" between IOTA code and IOTA Lib code...

When the IOTA receives something like this in its southbound interface:

```
p|52
```

It is the responsability of the agent to grab the measure, check that "52" and cast it to a number before pasing it to the library. The library will get a 52 (number), as shown in the `values` used to feed the test.

Is my interpretation correct?

dcalvoalonso replied:
Let's going to see the workflow with a real example: iotagent-ul with MQTT binding.
1. A device publishes an active attribute in a topic as you suggest: 
```
p|52
```

2. The iotagent-ul applies the following handle from [commonBindings.js](https://github.com/telefonicaid/iotagent-ul/blob/master/lib/commonBindings.js):
```
function singleMeasure(apiKey, attribute, device, message) {
    var values;
    values = [
        {
            name: attribute,
            type: guessType(attribute, device),
            value: message.toString()
        }
    ];

    iotAgentLib.update(device.name, device.type, '', values, device, function(error) {
        if (error) {
            config.getLogger().error(context,
                ' MEASURES-002: Couldn\'t send the updated values to the Context Broker due to an error: %s', error);
        } else {
            config.getLogger().debug(context, 'Single measure for device [%s] with apiKey [%s] successfully updated',
                device.id, apiKey);
        }
    });
}
```

3. As you can see this agent does not apply any kind of checking or cast regarding the input value of the measures and rely on the IOTA Lib code. It just tries to convert the soutbound data model to NGSI, e.g. in this case from UL to NGSI. From this point, we are exactly in the same point that we have with the unit tests of the IOTA Lib code.
4. If there is not a plugin activated, the IOTA Lib code just sends the attributes as they are received, so if an attribute with an integer value is received for NGSIv1 version, it would be sent to the CB. The CB with NGSIv1 for creation/update operations can use numbers, bool,
etc. but at the end they are transformed to strings and stored in that way internally so it's not a problem.
5. If there the expression plugin is enabled, the input value is passed to the expression parser which tries to apply the expression using the value as the input parameter.
6. The output value of the expression parser is used to replace the input one using replace instruction which causes a cast to String type.

fgalan replied:
Thanks for the detailed workflow! I have a doubt regarding

> As you can see this agent does not apply any kind of checking or cast regarding the input value of the measures and rely on the IOTA Lib code.

However, the value seems to be casted to string in the IOTA code:

```
value: message.toString()
```

Thus, for the IOTAgent `p|52` means always "52" (string) and never 52 (number).

(Maybe by "agent does not apply any kind of checking or cast" you mean what I refer as "casted to string"  Just clarifying... :)

Is this correct?

dcalvoalonso replied:
Upsssss!! You are completely right, I missed that specific and relevant instruction because I was just trying to answer quickly. As you point, iotagent-ul is doing the cast. Anyway, I guess that it is not mandatory because in NGSIv1 is possible to use numbers, bools... although when they are stored by the CB, they are transformed to Strings,

For instance, in the LWM2M (commonLwm2m.js)[https://github.com/telefonicaid/lightweightm2m-iotagent/blob/master/lib/services/lwm2mHandlers/commonLwm2m.js]:

```
function activeDataHandler(registeredDevice, name, type, value) {
    var attributes = [{
        name: name,
        type: type,
        value: value
    }];

    logger.debug(context, 'Handling data from device [%s]', registeredDevice.id);

    iotAgentLib.update(
        registeredDevice.name,
        registeredDevice.type,
        '',
        attributes,
        registeredDevice,
        function handleUpdateEntity(error) {
            if (error) {
                logger.error(context, 'Unknown error connecting with the Context Broker: ' + error);
            } else {
                logger.debug(context, 'Data handled successfully');
            }
        }
    );
}
```

fgalan replied:
So it seems that at the end the way in which the IOTAgent processes the attribute values before passing the to the IOTA Lib is a key point.

This has a important "coloraly": in order to support NGSIv2 in IOTAs it is not enough to support only the natives types allowed by NGSIv2 in the IOTA Lib. The code of the different IOTAs using the library should be reviewed, e.g. IOTA-UL shouldn't do unconditional cast to string. 

Maybe an issue about "NGSIv2 migration" should be opened in each one of the target IOTAs (IOTA-UL and IOTA-JSON at least).

fgalan replied:
Another doubt...

> The output value of the expression parser is used to replace the input one using replace instruction which causes a cast to String type.

If I understand correctly, you mean that if an expression is used, no matter the input of such expression, the output is always cast to String.

However, looking to your test updateContextExpressionPlugin1.json:

```
{
  "pressure": {
    "value": 1040,
    "type": "Integer"
  }
}
```

This is the result of applying the following

```
{
   object_id: 'p',
   name: 'pressure',
   type: 'Integer',
   expression: '${@pressure * 20}'
}
```

to value 52.

Should I understand that is `type` which defines which output type to use? I mean, if I use:


```
{
   object_id: 'p',
   name: 'pressure',
   type: 'String',
   expression: '${@pressure * 20}'
}
```

I'd get:

```
{
  "pressure": {
    "value": "1040",
    "type": "String"
  }
}
```

Is that correct?

dcalvoalonso replied:
In NGSIv1, no matter the input of such expression, the output always cast to String. But now, with the modifications that we have introduced to support JSON native types, you'r right `type` defines which output type to use.

To sum up:

* Input attribute:
```
{
  "pressure": {
    "value": 52,
    "type": "Integer"
  }
}
```

* Expression:

```
{
   object_id: 'p',
   name: 'pressure',
   type: 'Integer',
   expression: '${@pressure * 20}'
}
```

* Transformed attribute with previous version of IOTA Lib:
```
{
  "pressure": {
    "value": "1040",
    "type": "Integer"
  }
}
```

* Transformed attribute with current version of IOTA Lib (supporting NGSIv2)
```
{
  "pressure": {
    "value": 1040,
    "type": "Integer"
  }
}
```

dcalvoalonso replied:
Regarding:

> This has a important "coloraly": in order to support NGSIv2 in IOTAs it is not enough to support only the natives types allowed by NGSIv2 in the IOTA Lib. The code of the different IOTAs using the library should be reviewed, e.g. IOTA-UL shouldn't do unconditional cast to string.

> Maybe an issue about "NGSIv2 migration" should be opened in each one of the target IOTAs (IOTA-UL and IOTA-JSON at least)

I fully agree with you, we should open these issues since modifications will be needed in all the IOTAs.

fgalan replied:
Issues opened at:

* https://github.com/telefonicaid/iotagent-json/issues/250
* https://github.com/telefonicaid/iotagent-ul/issues/233
* https://github.com/telefonicaid/lightweightm2m-iotagent/issues/104

fgalan replied:
I'm looking in detail de documentation and I read (with regards to NGSIv1):

> Expressions can have two return types: String or Number. This return type must be configured for each attribute that is going to be converted. Default value type is String.
>
> Whenever a expression is executed without error, its result will be cast to the configured type. If the conversion fails (e.g.: if the expression is null or a String and is casted to Number), the measurement update will fail, and an error will be reported to the device.

The "its result will be cast to the configured type" contradicts "In NGSIv1, no matter the input of such expression, the output always cast to String". Is the documentation wrong? In that case, it should be fixed.

If the documentation is ok, it should clarify which type is used: the one defined for the attribute in the API or the one to which the expression belongs (I mean, if the expression belong to "string operations" the type is String and if the expression belongs to "numeric operations" the type is Number).

dcalvoalonso replied:
That part of the original documentation is correct... but only for the parse function of (expressionParser.js)[https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js]. You can check in [expression-test.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/test/unit/expressions/expression-test.js), that this function complies with

> Whenever a expression is executed without error, its result will be cast to the configured type. If the conversion fails (e.g.: if the expression is null or a String and is casted to Number), the measurement update will fail, and an error will be reported to the device.

But how this function is used by [expressionPlugin.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionPlugin.js):
1. In line 60, for the entity, the function processExpressionAttributes of (expressionParser.js)[https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js] is called.
2. processExpressionAttributes (line 179) function calls to expressionApplier
3. expressionApplier (line 157) function calls to applyExpression
4. applyExpression (line 145) function. Calls to processExpression and uses replace instruction to update the original value. As you can see in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace, the replace() method returns a new string. Please find below the code of this function:

```
function applyExpression(expression, context, typeInformation) {
    var expressionList = expression.match(/\$\{.*?\}/g) || [],
        substitutions = expressionList.map(processExpression(context)),
        expressionResult = expression;

    for (var i = 0; i < substitutions.length; i++) {
        expressionResult = expressionResult.replace(substitutions[i].original, substitutions[i].value);
    }

    return expressionResult;
}
```

5. processExpression (line 131) function. Calls to parse function BUT it uses as argument the type 'String'. Therefore, the results of the expression will be always cast to String.

Do you see my point?? Sorry, maybe I am not explaining well how this is implemented but it's a bit tricky....

fgalan replied:
Looking to expression-test.js file, it seems that the parse() function allows to define the output type. There is a function parameter for that, e.g.:

```
expressionParser.parse(arithmeticExpr[0], scope, 'Number', function(error, result)
...
expressionParser.parse('"Pruebas" # "DeStrings"', scope, 'String', function(error, result)
```

Next, looking to the logic used by the expressionPlugin.js (the one you perfectly describe in the 5 steps) I understand that the "link" with parse() is in the following point (https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js#L136, processExpression() function):

```
 result = parse(cleanedExpression, context, 'String');
```

So it seems that although we have a poweful function able to cast to Numbers and String we are capping ourselves to use only the String mode :)

If my interpretation is correct (please confirm... as you say this is tricky so let's confirm common understading in each step in the dicussion :) I'd suggest the following modifications:

1. Add a comment just above L136 in order to explain that the function is able to use String and Number but we are using only String. Suggestion (please freely edit the text):

```
// Note that parse() function allows both String way processing and Number way processing
// (have a look to  expression-test.js). However, here we only use one of the possibilities, i.e. String
```

2. Adjust https://github.com/dcalvoalonso/iotagent-node-lib/blob/master/doc/expressionLanguage.md#-types section given that it is confusing (or even wrong). Suggestion (please review if I'm right in the last paragraph):

> The way the `parse()` function works (at expressionParser.js) is as follows:
>
> * Expressions can have two return types: String or Number. This return type must be configured for each attribute that is going to be converted. Default value type is String.
> * Whenever a expression is executed without error, its result will be cast to the configured type. If the conversion fails (e.g.: if the expression is null or a String and is casted to Number), the measurement update will fail, and an error will be reported to the device.
> 
> However, the usage that the Transformation Expression does of that function is using always String type. That means that at the end the resulf of the expression will be always casted to String. However, in NGSIv2 that String result could be re-cast to the right type (i.e. the one defined for the attribute in the provision operation). Have a look to the [NGSIv2 section] for more information on this.

dcalvoalonso replied:
Your analysis is correct so I guess that we have now a common understanding of this plugin! :)

I will update the code to include your suggestions. Thanks!!!

dcalvoalonso replied:
Fixed in 45d40dd86730117e13f4e62b5dd8191e51b5bc6f

fgalan commented on lib/plugins/expressionParser.js:
```diff
@@ -163,6 +165,19 @@ function expressionApplier(context, typeInformation) {
+        }
+        else if (attribute.type === 'Null') {
+            newAttribute.value = null;
+        }
```
expressionLanguage.md should include information about how the mapping works in NGSIv2, i.e. how the different types (some of them news, e.g. Boolean, etc.) work. Maybe the best way by the time beinb would be a separate section (although in the future, if NGSIv1 is removed from IOTA library the entire .md should be rewritten).

dcalvoalonso replied:
Fixed in d6581fc4915c900b66c0144df3f7d19bd11253b2

fgalan replied:
I'm looking in detail de documentation and I read (with regards to NGSIv1):

> Expressions can have two return types: String or Number. This return type must be configured for each attribute that is going to be converted. Default value type is String.
>
> Whenever a expression is executed without error, its result will be cast to the configured type. If the conversion fails (e.g.: if the expression is null or a String and is casted to Number), the measurement update will fail, and an error will be reported to the device.

The "its result will be cast to the configured type" contradicts "In NGSIv1, no matter the input of such expression, the output always cast to String". Is the documentation wrong? In that case, it should be fixed.

If the documentation is ok, it should clarify which type is used: the one defined for the attribute in the API or the one to which the expression belongs (I mean, if the expression belong to "string operations" the type is String and if the expression belongs to "numeric operations" the type is Number).

dcalvoalonso committed:
Unit tests are reordered and commented for further clarification

dcalvoalonso committed:
Description of NGSIv2 types for expressionLanguage.md

dcalvoalonso committed:
Correct markdown in doc/expressionLanguage.md

fgalan commented on doc/expressionLanguage.md:
```diff
@@ -175,3 +176,35 @@ The following table shows expressions and their expected outcomes for a measure
+
+As it is explained in previous sections, expressions can have two return types: String or Number, being the former one the default. Whenever a expression is executed without error, its result will be cast to the configured type. 
+
+On one hand, in NGSIv1 since all attributes'values are string, in the expression parser the expression type is set always to String and the transformation of the information coming from the SouthBound is done using replace instruction. Therefore, values sent to the CB will always be Strings. This can be seen in (#execution) example.
```
Typo: attributes'values

dcalvoalonso replied:
Fixed in 776fe93a154bdd014f5bb7b37c8e165a52ab66d1

fgalan commented on doc/expressionLanguage.md:
````diff
@@ -175,3 +176,35 @@ The following table shows expressions and their expected outcomes for a measure
+status: true
+```
+
+The value true will be sent to the Context Broker since true will be converted to 1 before applying the expression and its reult (20) will be converted back to true (Everything With a "Value" is True in javacript).
````
Typo: reult -> result

dcalvoalonso replied:
Fixed in 776fe93a154bdd014f5bb7b37c8e165a52ab66d1

fgalan commented on doc/expressionLanguage.md:
```diff
@@ -175,3 +176,35 @@ The following table shows expressions and their expected outcomes for a measure
+On one hand, in NGSIv1 since all attributes'values are string, in the expression parser the expression type is set always to String and the transformation of the information coming from the SouthBound is done using replace instruction. Therefore, values sent to the CB will always be Strings. This can be seen in (#execution) example.
+
+On the other hand, NGSIv2 fully supports all the types described in the JSON
+specification (string, number, boolean, object, array and null). Therefore, the result of an expression must be cast to the appropriate type in order to avoid inconsistencies between the type field for an attribute and the type of the value that is being sent.
```
Do "the appropiate type" refers to the type used to define the attribute? I mean, in the example shown in this documentation:

* The "appropiate type" for `location` would be String (actually it uses geo:point, but it is of string nature)
* The "appropiate type" for `temperature` and `fillingLevel` would be Number

or maybe "appropiate type" refers to the one associated to the expression being applied (so the type used to define the attribute is ignored in this case)?

fgalan replied:
(This is actually related with this other comment https://github.com/dcalvoalonso/iotagent-node-lib/pull/3#discussion_r154142079... I'm just trying to fit all the pieces together in my mind and, in the process, trying to improve the tests and documentation :)

dcalvoalonso replied:
Fixed in 776fe93a154bdd014f5bb7b37c8e165a52ab66d1

dcalvoalonso committed:
Improves NGSIv2 section of expressionLanguage doc

dcalvoalonso committed:
Passing LINT

fgalan commented on doc/expressionLanguage.md:
```diff
@@ -175,3 +176,35 @@ The following table shows expressions and their expected outcomes for a measure
+
+As it is explained in previous sections, expressions can have two return types: String or Number, being the former one the default. Whenever an expression is executed without error, its result will be cast to the configured type. 
+
+On one hand, in NGSIv1 since all attributes' values are of type String, in the expression parser the expression type is set always to String and the transformation of the information coming from the SouthBound is done using replace instruction. Therefore, values sent to the CB will always be Strings. This can be seen in (#execution) example.
```
Typo: #execution (it is not using the right syntax for links in markdown).

fgalan replied:
Fixed in 4026d75714dd6ef51230bd01c8482c0a89e36923

fgalan commented on doc/expressionLanguage.md:
```diff
@@ -175,3 +176,35 @@ The following table shows expressions and their expected outcomes for a measure
+
+1. Variables will be cast to Number or String depending on the expression type.
+2. The expression will be applied
+3. The output type will be cast again to the original attribute type.
```
As sub-bullets of 3, I would suggest to add the following text in order to be completly splicit about how the cast is done (please, check if it correct... I could have understood something incorrectly):

* If attribute type is "Integer" then the value is casted to integer (JSON number)
* If attribute type is "Float" then the value is casted to float (JSON number)
* If attribute type is "Boolean" then the value is cast to boolean (JSON boolean)
* If attribute type is "Null" then the value is cast to null (JSON null)

dcalvoalonso replied:
Fixed in 4026d75714dd6ef51230bd01c8482c0a89e36923

fgalan commented on doc/expressionLanguage.md:
````diff
@@ -175,3 +176,35 @@ The following table shows expressions and their expected outcomes for a measure
+status: true
+```
+
+The value true will be sent to the Context Broker since true will be converted to 1 before applying the expression and its result (20) will be converted back to true (Everything With a "Value" is True in Javascript).
````
I'd suggest to refactor this text to follow the workflow you mention more explicit, step by step. I think it would be more illustrative. Something like this (feel free to reuse the text):

1. The expression `*` is a multiplication, so the expression type makes `status` to be casted to Number. The cast of `true` to number is 1 (everything with a "value" is true in JavasScript).
2. Expression is evaluated, resulting in 20
3. The attribute type is `Boolean` so the result is casted to Boolean before sending it to CB. The cast of `20` to boolean is 20 (everything with a "value" is true in JavasScript).

dcalvoalonso replied:
Fixed in 4026d75714dd6ef51230bd01c8482c0a89e36923

fgalan commented on lib/plugins/expressionParser.js:
```diff
@@ -163,6 +165,19 @@ function expressionApplier(context, typeInformation) {
+        else if (attribute.type === 'Boolean') {
+            newAttribute.value = (newAttribute.value === true);
+        }
+        else if (attribute.type === 'Null') {
```
NGSIv2 uses "None" for this (see http://telefonicaid.github.io/fiware-orion/api/v2/stable/ section "Partial representations". Could make sense to change to align?

dcalvoalonso replied:
Fixed in 4026d75714dd6ef51230bd01c8482c0a89e36923

fgalan commented on test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js:
```diff
@@ -0,0 +1,774 @@
+        beforeEach(function() {
+            nock.cleanAll();
+
+            contextBrokerMock = nock('http://192.168.1.1:1026')
```
Probably in this case you don't need the mock, as the fail this test is testing avoids any update to CB, isn't it?

dcalvoalonso replied:
Fixed in 4026d75714dd6ef51230bd01c8482c0a89e36923

fgalan commented on test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js:
```diff
@@ -0,0 +1,774 @@
+                        object_id: 'e',
+                        name: 'consumption',
+                        type: 'Float',
+                        expression: '${trim(@spaces)}'
```
There isn't any attribute named `spaces`... maybe I'm missing something?

dcalvoalonso replied:
Fixed in 4026d75714dd6ef51230bd01c8482c0a89e36923

fgalan commented on test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js:
```diff
@@ -0,0 +1,774 @@
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', 'gardens')
+                .post('/v2/entities/ws1/attrs', utils.readExampleFile(
+                    './test/unit/ngsiv2/examples/contextRequests/updateContextExpressionPlugin12.json'))
```
The definition for e in WeatherStationMultiple is:

```
{
                        object_id: 'e',
                        name: 'consumption',
                        type: 'Float',
                        expression: '${trim(@spaces)}'
}
```

If there isn't any `spaces` attribute, then `${trim(@spaces)}` results in `""` empty string. However, updateContextExpressionPlugin12.json shows:

```
  "consumption": {
    "type": "Float",
    "value": 0.44
  }
```

It is like the expression were being ignored and the attribute value passed directly to CB...

dcalvoalonso replied:
Fixed in 4026d75714dd6ef51230bd01c8482c0a89e36923

fgalan replied:
Considering the code previous to the fix, I understand that `${trim(@spaces)}` should evaluate to `""` (as `spaces` attributes doesn't exist), which cast to Float I understand is `0.0`. Thus, why we had (previous to the fix)?

```
  "consumption": {
    "type": "Float",
    "value": 0.44
  }
```

instead to

```
"consumption": {
    "type": "Float",
    "value": 0.0
  }
```

Although this has been fixed changing expression `'${trim(@updated)}'` it would be great to know how it worked before. That behaviour could hide some kind of bug...

dcalvoalonso replied:
This worked because the plugin checks if attribute name matches the expression variables before applying it.

In 179 of [expressionParser.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js):
* list argument contains the active attributes provisioned for the device. These attributes may include expressions.
* context contains the measure received from the Southbound, e.g., {"consumption":0.44}
* contextAvailable function (line 170) extracts the variables from the expression and returns true if they match the measure. In this example, the comparison of line 176 is between ["spaces"] and ["consumption"]. Therefore, the expression is not applied and the IOTA will send to the CB an updateContext query that contains directly the NGSI translation of the Southbound measure.

fgalan replied:
Ok, now I understand how it works more or less. If that behaviour is not explained in the documentation (I haven't find it, but I could be wrong) I'd suggest to add the following explanatory text to expressionLanguage.md (maybe at then end of "Expression execution" section):

> Note that expressions are only applied if the attribute name (as recevied by the IOT Agent in the southbound interface) matches the expression variable. Otherwise the southbound value is used directly. Let's illustrate with the following example:
> ```
>   "consumption": {
>    "type": "Float",
>    "value": "${trim(@spaces)}"
>  }
> ````
>
> Case 1: the following measure is received at southbound interface:
> ```
> consumption: 0.44
> ```
> as `spaces` attribute is not included, then the expression is not applied and the
`consumption` measure value is directly used, so the following is sent to CB:
> ``` 
>"consumption": {
>  "type": "Float",
>  "value": 0.44
> }
> ```
> Case 2: the following measure is recieved at southbound interface:
>
> ```
> consumption: 0.44
> spaces: "  foobar  "
>```
> as `spaces` is included, then the expression is evaluated, so overriding the 0.44 value and sending the following to CB:
>
> ``` 
>"consumption": {
>  "type": "Float",
>  "value": "foobar"
> }
> ```

Please, ensure case 2 is as I describe. I could be wrong. Maybe it is a good idea to include a test case to ensure it.

dcalvoalonso replied:
Fixed in 3c99f7f51aa6950661470e0a88525764306654c9

dcalvoalonso replied:
I have added the example you proposed with just one minor change:
```
  "consumption": {
   "type": "String",
   "value": "${trim(@spaces)}"
 }
````

As you can see I have used String type instead of Float. If we used a Float type, at the end "foobar" would be cast to Float, getting null value.

dcalvoalonso replied:
In my eyes, it is not necessary to add new unit tests. These cases are already covered in several of the already existing tests, see the lines 232, 267 and 303.

fgalan replied:
As far as I understand, the test cases you refer (L232, L267 and L303) use a pattern in which the expression includes the attribute or no expression is used, e.g:

```
{
                        object_id: 'p',
                        name: 'pressure',
                        type: 'Integer',
                        expression: '${@pressure * / 20}'
},
...
{
                        object_id: 'h',
                        name: 'humidity',
                        type: 'Percentage'
},
...
```

What I mean is a pattern in which the expression includes a *different* attribute. E.g.:

```
{
                        object_id: 'x',
                        name: 'consumption_x',
                        type: 'Float',
                        expression: '${@pressure * 20}'
},
```

Case 1: as test input use:

```

        var values = [
            {
                name: 'x',
                type: 'Float',
                value: 0.44
            }
];
```

And get as output:

```
"consumption_X": {
   "type": "Float",
   "value": 0.44
 }
```

Case 2: as test input use:

```
        var values = [
            {
                name: 'p'',
                type: 'Integer',
                value: 10
            }
];
```

And get as output (expression overrides "direct" value):

```
"consumption_x": {
   "type": "Float",
   "value": 200
 }
```

Case 3: as test input use:

```
        var values = [
            {
                name: 'x',
                type: 'Float',
                value: 0.44
            },
            {
                name: 'p'',
                type: 'Integer',
                value: 10
            }
];
```
(Same as case 2, but with an "overriding" situation)

And get as output:

```
"consumption_x": {
   "type": "Float",
   "value": 200
 }
```

What do you think?

dcalvoalonso replied:
Ok, now I get your point! I will add these new tests. :)

dcalvoalonso replied:
Fixed in e3bc553d7598e699abe68f22bc4da9bbb3c6cf85

fgalan commented on test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js:
```diff
@@ -0,0 +1,774 @@
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', 'gardens')
+                .post('/v2/entities/ws1/attrs', utils.readExampleFile(
+                    './test/unit/ngsiv2/examples/contextRequests/updateContextExpressionPlugin10.json'))
```
```
update: true -> ${@update * 20} -> ${ true * 20 } -> $ { true } -> true
```

However updateContextExpressionPlugin10.json shows false. Which step in my sequence is wrong?

dcalvoalonso replied:
update: true -> ${@update * 20} -> ${ 1 * 20 } -> $ { 20 } -> $ { "20"} -> False 

Arithmetic expression so true is cast to integer ('1'). You are missing the cast to String at the end of the expression.

fgalan replied:
If I'm understanding correctly, "20" (string) is casted to false (boolean) in JavaScript. However, in some point of the documentation it is said "everything with a "value" is true in JavasScript"... "20" is a "value", isn't it? Or maybe I'm missing something? :)

dcalvoalonso replied:
Finally in order to do the final cast to boolean, we are using the following:

```
newAttribute.value = (newAttribute.value === 'true' || newAttribute.value === '1');
```

This implies that only 'true' or '1' String values will be converted to true. Please note that newAttribute.value will be always a String due to the use of the replace instruction.
The documentation was not correct because my initial idea was to use directly the [Boolean function] (https://www.w3schools.com/js/js_booleans.asp), but in this case even 'false' will be true, because

> Everything With a "Value" is True

dcalvoalonso replied:
Fixed in 65827ac053ce095effbb2d1fdf459eb596f38d27

dcalvoalonso committed:
Address comments in PR #3

dcalvoalonso committed:
Improvind docs for expressionLanguage

dcalvoalonso committed:
Expression parser. Boolean type

dcalvoalonso committed:
Cast from String to Boolean

fgalan commented on test/unit/ngsiv2/examples/contextRequests/updateContextExpressionPlugin3.json:
```diff
@@ -0,0 +1,8 @@
+{
```
I understand that after removing the mock at 4026d75714dd6ef51230bd01c8482c0a89e36923 this file is no longer needed.

Removing it would create a "gap" in the numbering so the whole serie of updateContextExpressionPluginX.json could be renumbered. However, this renumbering is a minor thing, not really needed.

dcalvoalonso replied:
Fixed in cb737a4b8cdbf5d0ebe51fd6b1bb69410c6a765b

fgalan commented on doc/expressionLanguage.md:
```diff
@@ -175,3 +176,52 @@ The following table shows expressions and their expected outcomes for a measure
+
+More examples of this workflow are presented below for the different types of attributes supported in NGSIv2 and the two possible types of expressions: Integer (arithmetic operations) or Strings.
+
+* pressure: 52 -> ${@pressure * 20} -> ${ 52 * 20 } -> $ { 1040 } -> $ { "1040"} -> 1040
```
Nice! This is *very* clarifying and uselful for plugin users :)

NTC

fgalan commented on package.json:
```diff
@@ -35,9 +35,10 @@
     "async": "1.5.2",
```
Is there any real change in the file? Or maybe the reordering is a side-effect of adding boolean package in one commit, removing in a next one?

(No need to change to make the file exactly at is was, this is just to confirm)

dcalvoalonso replied:
Nothing has changed actually, as you mentioned, it's just the result of adding and removing boolean package.

fgalan replied:
Ok. NTC

fgalan commented on doc/expressionLanguage.md:
```diff
@@ -175,3 +176,52 @@ The following table shows expressions and their expected outcomes for a measure
+3. 20 is cast to `20` since Expression Plugin always use String as Expression type.
+4. The attribute type is `Boolean` so the result is casted to Boolean before sending it to CB. The cast of `20` to boolean is false (only `true` or `1` are cast to true).
+
+More examples of this workflow are presented below for the different types of attributes supported in NGSIv2 and the two possible types of expressions: Integer (arithmetic operations) or Strings.
```
It would be clarifying to include the type of the attribute in each case, eg.:

pressure (of type "Integer): 52 -> ${trim(@pressure)} -> ${trim("52")} -> $ { "52" } -> $ { "52"} -> 52
...
update (of type "Boolean"): false -> ${@update * 20} -> ${ 0 * 20 } -> $ { 0 } -> $ { "0"} -> False

dcalvoalonso replied:
Fixed in cb737a4b8cdbf5d0ebe51fd6b1bb69410c6a765b

dcalvoalonso committed:
Addressing comments of PR #3

fgalan commented:
To sum up, I think at the present moment only this two issues are opened in this PR:

* https://github.com/dcalvoalonso/iotagent-node-lib/pull/3#discussion_r154682873, involving documentation changes and code comments
* https://github.com/dcalvoalonso/iotagent-node-lib/pull/3#discussion_r154686469, a "weird" behaviour (at least from my point of view) that would need some research and clarification

Anything else I may have missed? :)

dcalvoalonso committed:
Addressing comments of PR #3

fgalan commented on doc/expressionLanguage.md:
```diff
@@ -110,12 +111,12 @@ The only expression rule that will be executed will be that of the 'fillingLevel
+- Expressions can have two return types: String or Number. This return type must be configured for each attribute that is going to be converted. Default value type is String.
+- Whenever an expression is executed without error, its result will be cast to the configured type. If the conversion fails (e.g.: if the expression is null or a String and is cast to Number), the measurement update will fail, and an error will be reported to the device.
+
+However, the usage that the Expression Translation plugin does of that function is using always String type. That means that at the end, the result of the expression will be always cast to String. However, in NGSIv2 that String result could be re-cast to the right type (i.e. the one defined for the attribute in the provision operation). Have a look at the [NGSIv2 section] for more information on  ...
```
Typo: [NGSIv2 section] should link to the actual section.

dcalvoalonso replied:
Fixed in 626e11f6c1556acad8c2ff6513a06ffbe7e10fa1

dcalvoalonso replied:
With my last comment and the update of the docs, I think that the last two issues are solved.

dcalvoalonso committed:
Update expressionLanguage.md

dcalvoalonso committed:
Update expressionLanguage.md

fgalan commented on doc/expressionLanguage.md:
````diff
@@ -106,16 +107,51 @@ level: 85.3
+
+- Case 1: the following measure is received at the southbound interface:
+```
+consumption: "0.44"
````
L120 uses "0.44" but L133 uses 0.44. Although both could be ok (type "String" will cast both them correctly at CB interaction) I'd suggest to use the same in both cases.

dcalvoalonso replied:
Fixed in 27e54d40bca4fcf8961987beea741a1778f9791b

fgalan commented on doc/expressionLanguage.md:
````diff
@@ -106,16 +107,51 @@ level: 85.3
+spaces: "  foobar  "
+```
+
+As `spaces` attribute is included, then the expression is evaluated, so overriding the 0.44 value and sending the following to CB:
````
0.44 -> "0.44" (as type is String, it would make more sense)

dcalvoalonso replied:
Fixed in 27e54d40bca4fcf8961987beea741a1778f9791b

fgalan commented on doc/expressionLanguage.md:
```diff
@@ -175,3 +211,52 @@ The following table shows expressions and their expected outcomes for a measure
+
+Currently, the expression parser does not support JSON Arrays and JSON document. A new issue has been created to address this aspect https://github.com/telefonicaid/iotagent-node-lib/issues/568. For the rest of types the workflow will be the following:
+
+1. Variables will be cast to Number or String depending on the expression type.
```
In a second read and taking into account what we added recently (*That means that at the end, the result of the expression will be always cast to String*) if I'm interpreting it correctly, I think this step should be changed to:

```
1. Variables will be cast to String no matter the expression type (see [comments above](#types) regarding this)
```

dcalvoalonso replied:
Fixed in 27e54d40bca4fcf8961987beea741a1778f9791b

fgalan commented on doc/expressionLanguage.md:
````diff
@@ -175,3 +211,52 @@ The following table shows expressions and their expected outcomes for a measure
+status: true
+```
+
+1. The expression `*` is a multiplication, so the expression type makes `status` to be casted to Number. The cast of `true` to number is 1 (everything with a "value" is true in JavasScript).
````
In fact I undesrstand *"(everything with a "value" is true in JavasScript)"* explains the other way of the cast (from number to bool)... Maybe that text between brackets should be simply removed.

dcalvoalonso replied:
Fixed in 27e54d40bca4fcf8961987beea741a1778f9791b

dcalvoalonso committed:
Update expressionLanguage.md

fgalan commented on test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js:
```diff
@@ -0,0 +1,763 @@
+        });
+    });
+
+    describe('When an update comes for attributes with numeric expressions and type integer', function() {
```
Test case L364-L392 seems to be equal to test case L303-L332. Maybe I'm missing some difference?

dcalvoalonso replied:
Fixed in 799b13427fa60951e8b67e6fb1d1df0e8402a89f

dcalvoalonso replied:
I suppose that when we added more tests to cover all the possible cases, we duplicated this one.

dcalvoalonso committed:
Addressing comments of PR#3

dcalvoalonso committed:
Merge branch 'task/expressionPluginNgsi2' of https://github.com/dcalvoalonso/iotagent-node-lib into task/expressionPluginNgsi2

dcalvoalonso committed:
Addressing comments of PR#3

fgalan commented on test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js:
```diff
@@ -0,0 +1,833 @@
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', 'gardens')
+                .post('/v2/entities/light1/attrs', utils.readExampleFile(
+                    './test/unit/ngsiv2/examples/contextRequests/updateContextExpressionPlugin13.json'))
```
It seems that updateContextExpressionPlugin13.json is not part of the PR... Maybe you missed `git add` of that file before your last commit.

dcalvoalonso replied:
Fixed in f50c6f43594efa5bd5122fd2c4c15bbbbe7c405d

dcalvoalonso committed:
Adding missing file

fgalan commented:
LGTM

Thanks for the contribution! This time has taken more time than usual, due to the several iterations and the many comments. However, this is normal, as expressions plugin is probably the one with most complex functionality and have several corner cases and special situations, not all originally documented.

Passing the ball to @AlvaroVega for final LGTM. Once he provides it, please "Stash&Merge" the PR.

dcalvoalonso commented:
Thanks for your feedback @fgalan , your comments have been very constructive and I think that together we have achieved a good understanding of this complex plugin.

@AlvaroVega, what do you think about this PR?

fgalan commented:
For this particular case if @AlvaroVega LGTM takes time to come I think you can "Stash&Merge" based in my single LGTM. This PR has been reviewed in deep and probably with one LGTM is enough.

dcalvoalonso commented:
Ok, let's going to wait until Monday morning and if he does not say anything, I will proceed with the "Stash&Merge".

Thanks again!!
